samza-kafka/src/main/scala/org/apache/samza/system/kafka/NewKafkaSystemConsumer.java

```diff
@@ -0,0 +1,412 @@
+      proxy.stop(TimeUnit.MINUTES.toMillis(5));
+    }
+
+    try {
+      synchronized (kafkaConsumer) {
+        LOG.info("Closing kafka consumer " + kafkaConsumer);
+        kafkaConsumer.close();
+      }
+    } catch (Exception e) {
+      LOG.warn("failed to stop SamzaRawKafkaConsumer + " + this, e);
+    }
+  }
+
+  /*
+   record the ssp and the offset. Do not submit it to the consumer yet.
+   */
+  @Override
+  public void register(SystemStreamPartition systemStreamPartition, String offset) {
+    if (started.get()) {
+      String msg =
+          String.format("Trying to register partition after consumer has been started. sn=%s, ssp=%s", systemName,
+              systemStreamPartition);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    if (!systemStreamPartition.getSystem().equals(systemName)) {
+      LOG.warn("ignoring SSP " + systemStreamPartition + ", because this consumer's system is " + systemName);
+      return;
+    }
+    super.register(systemStreamPartition, offset);
+
+    TopicPartition tp = toTopicPartition(systemStreamPartition);
+
+    topicPartitions2SSP.put(tp, systemStreamPartition);
+
+    LOG.info("Registering ssp = " + systemStreamPartition + " with offset " + offset);
+
+    String existingOffset = topicPartitions2Offset.get(tp);
+    // register the older (of the two) offset in the consumer, to guarantee we do not miss any messages.
+    if (existingOffset == null || compareOffsets(existingOffset, offset) > 0) {
+      topicPartitions2Offset.put(tp, offset);
+    }
+
+    samzaConsumerMetrics.registerTopicAndPartition(toTopicAndPartition(tp));
+  }
+
+  /**
+   * Compare two String offsets.
+   * Note. There is a method in KafkaAdmin that does that, but that would require instantiation of systemadmin for each consumer.
+   * @param off1
+   * @param off2
+   * @return see {@link Long#compareTo(Long)}
+   */
+  public static int compareOffsets(String off1, String off2) {
+    return Long.valueOf(off1).compareTo(Long.valueOf(off2));
+  }
+
+  @Override
+  public String toString() {
+    return systemName + "/" + clientId + "/" + super.toString();
+  }
+
+  @Override
+  public Map<SystemStreamPartition, List<IncomingMessageEnvelope>> poll(
+      Set<SystemStreamPartition> systemStreamPartitions, long timeout) throws InterruptedException {
+
+    // check if the proxy is running
+    if (!proxy.isRunning()) {
+      stop();
+      if (proxy.getFailureCause() != null) {
+        String message = "KafkaConsumerProxy has stopped";
+        throw new SamzaException(message, proxy.getFailureCause());
+      } else {
+        LOG.warn("Failure cause is not populated for KafkaConsumerProxy");
+        throw new SamzaException("KafkaConsumerProxy has stopped");
+      }
+    }
+
+    Map<SystemStreamPartition, List<IncomingMessageEnvelope>> res = super.poll(systemStreamPartitions, timeout);
+    return res;
+  }
+
+  public static TopicAndPartition toTopicAndPartition(TopicPartition tp) {
```

Comment:
All of the following static methods(`toTopicAndPartition(topicParittion)`, `toTopicAndPartition(SystemStreamPartition`)  here seems to be named with similar names and have different parameter type.  Do we need a static method here just to create a object. Can we instead inline the method invocations?